### DISCUSSION TITLE
cbuffer array

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#256] [Q&A] [answered]
What would be the procedure for uploading data to a constant/static array of data in a cbuffer? I set things up the way I expected it should work, and I get no errors or warnings, but the data does not get uploaded to the variable being used by the shaders. Is there a sample or example code that uploads data to a static cbuffer array?

This is what my shader definition looks like:
```
cbuffer POISSONSAMPLES
{
	float4 PoissonSamp[128];
};
```

I have this defined in a header that is only visible to the buffer uploader:
```
const vec4 POISSONSAMPLES[] = {
	{ -0.9406119,	0.2160107,	0.9650965,	0.4978694 },
	{ -0.9200030,	0.0313576,	0.9205372,	0.1564050 },
	{ -0.9178760,	-0.2841548,	0.9608539,	0.3357979 },
	...
};
```

Then this is what my buffer creation/upload looks like:
```
// prepare
Diligent::BufferDesc bdesc;

// sync debugging name (static exe string)
bdesc.Name = "poisson";

// requested size & stride
bdesc.Size = sizeof(vec4)*128;
bdesc.ElementByteStride = bdesc.Size;

// properties
bdesc.BindFlags = Diligent::BIND_FLAGS::BIND_SHADER_RESOURCE;
bdesc.Usage = Diligent::USAGE::USAGE_IMMUTABLE;
bdesc.CPUAccessFlags = Diligent::CPU_ACCESS_FLAGS::CPU_ACCESS_NONE;
bdesc.Mode = Diligent::BUFFER_MODE::BUFFER_MODE_STRUCTURED;
bdesc.MiscFlags = Diligent::MISC_BUFFER_FLAGS::MISC_BUFFER_FLAG_NONE;
bdesc.ImmediateContextMask = bit(Context.GetDesc().ContextId);

// prepare buffer data
Diligent::BufferData bdata( POISSONSAMPLES, sizeof(vec4)*128 );
	
// create the buffer
Device.CreateBuffer( bdesc, &bdata, &Inst );

// transition state
Diligent::RESOURCE_STATE old_state = Inst.GetState();
if( old_state != Diligent::RESOURCE_STATE_SHADER_RESOURCE )
{
	// create state transition barrier
	Diligent::StateTransitionDesc barrier( &Inst, old_state, Diligent::RESOURCE_STATE_SHADER_RESOURCE, Diligent::STATE_TRANSITION_FLAG_UPDATE_STATE );

	// transition the state
	Context.TransitionResourceStates( 1, &barrier );
}
```

Then finally this is what my pipeline resource signature entry looks like:

```
...
{ Diligent::SHADER_TYPE_VERTEX | Diligent::SHADER_TYPE_PIXEL, "POISSONSAMPLES", 1, Diligent::SHADER_RESOURCE_TYPE_CONSTANT_BUFFER, Diligent::SHADER_RESOURCE_VARIABLE_TYPE_STATIC, Diligent::PIPELINE_RESOURCE_FLAG_NO_DYNAMIC_BUFFERS },
...
```

I first tried uploading the data as an array of 128 float4 elements, but kept getting errors about it being an array. So I switched to trying to upload as a single element. Are there any obvious problems with this approach? Note that most of the code above is executed through utility functions in a much more generic way, but I translated it to very specific setup code to make it more clear, so hopefully I didn't make any serious blunders while doing so.

Really appreciate any help!

Edit: For some reason, the data upload works if I change the pipeline resource variable type from `SHADER_RESOURCE_VARIABLE_TYPE_STATIC` to `SHADER_RESOURCE_VARIABLE_TYPE_MUTABLE`. Its possible I'm just misunderstanding the meaning of static? Would there be a different way to upload data for a static variable?

**Comment by TheMostDiligent** [comment, marked as answer]
Variable type is orthogonal to buffer usage. Immutable buffers are initialized when they are created and your code looks correct.
Variable type defines how you are going to bind the resource to that variable.
Static variables are set directly through pipeline/signature.
Mutable and dynamic variables are set through SRB.

**Reply by RobertMtx** [reply]
Ok, thanks. Any ideas why the data upload would not be reflected when it is set to `SHADER_RESOURCE_VARIABLE_TYPE_STATIC`?

I also changed my `cbuffer` into `StructuredBuffer<float4> PoissonSamp;` and changed the resource signature from `Diligent::SHADER_RESOURCE_TYPE_CONSTANT_BUFFER` to `Diligent::SHADER_RESOURCE_TYPE_BUFFER_SRV`. It's possible that one of these changes is why its working. I guess I need to do some more testing to see if I can get it to work as a constant buffer. I read somewhere that it is faster for the GPU to read from a constant buffer than a structured buffer. Since this is data that will never change once its uploaded, I guess it would be best to set it as static?

Also, and I don't believe this has anything to do with Diligent Engine, but I'm a little unclear on the differences between `cbuffer` and `ConstantBuffer<struct>`. Is `ConstantBuffer<struct>` just a more up-to-date syntax of the same type of object? That's what I thought at first, but it looks like most or all of the Diligent Engine examples are using `cbuffer`.

Edit: As far as setting the variable through the signature, do you mean setting up an entry through the `Diligent::PipelineResourceDesc` array? If so, that's what I was referring to by the following line:
```
...
{ Diligent::SHADER_TYPE_VERTEX | Diligent::SHADER_TYPE_PIXEL, "POISSONSAMPLES", 1, Diligent::SHADER_RESOURCE_TYPE_CONSTANT_BUFFER, Diligent::SHADER_RESOURCE_VARIABLE_TYPE_STATIC, Diligent::PIPELINE_RESOURCE_FLAG_NO_DYNAMIC_BUFFERS },
...
```
But this doesn't associate it with the buffer, does it? Perhaps I'm missing something important here?

Edit 2: I just found `Diligent::IPipelineResourceSignature::GetStaticVariableByName()`. I'm assuming I need to use this interface to access and set the variable? I will dig deeper. Thanks a lot for the information.

**Reply by TheMostDiligent** [reply]
> Any ideas why the data upload would not be reflected when it is set to SHADER_RESOURCE_VARIABLE_TYPE_STATIC?

Again, data upload is orthogonal to which variable you bind it to. It is either uploaded to a buffer or not.

>  It's possible that one of these changes is why its working.

If this is what makes it work, then it is 100% possibility :)

> I read somewhere that it is faster for the GPU to read from a constant buffer than a structured buffer.

Yes, constant buffers are faster, but they are limited to 64 KB. Structured buffers have no limitation on the size.

> the differences between cbuffer and ConstantBuffer<struct>

`ConstantBuffer<struct>` is the modern way that is supported by DXC, but it is not supported by legacy D3DCompiler.

>  I just found Diligent::IPipelineResourceSignature::GetStaticVariableByName(). I'm assuming I need to use this interface to access and set the variable? I will dig deeper. Thanks a lot for the information.

If you are using static variables with pipeline resource signatures, then you bind the variable directly to signature. Diligent will complain if you don't bind a resource that is required by PSO.

**Reply by RobertMtx** [reply]
This was actually the first time I had used a static variable, so I didn't realize there would be situations where you could or would assign variables through the pipeline signature/state instead of the SRB. I get it now. Is there any significant advantage to using static variables over mutable variables? Or is it more of a dispersion thing?

**Reply by TheMostDiligent** [reply]
This is purely for convenience. If there is a resource that is shared between all instances of SRB created from pipeline/signature, you can bind it directly to PSO/Signature and when you create an SRB, it will automatically be bound.

**Reply by RobertMtx** [reply]
This explains why mine still wasn't working. I was setting the static variable after creating the SRB lol

**Reply by TheMostDiligent** [reply]
Diligent prints an error when draw command misses resources.

**Reply by RobertMtx** [reply]
Into the output window? That's usually where I watch for Diligent Engine warnings/errors. I wasn't getting any type of message about it, as far as I could tell. Just my buffer data was blank/zero. If you mean through an error handler, I haven't set one up yet. But if there are errors being sent to that which do not get sent to the output window, then I probably should right away.

**Reply by TheMostDiligent** [reply]
In the output, Diligent by default sends all messages there.
